expr/shaper.go

```diff
@@ -603,10 +626,44 @@ func castType(zctx *zson.Context, inType, specType zng.Type) (zng.Type, error) {
 			return zctx.LookupTypeArray(out), nil
 		}
 		return zctx.LookupTypeSet(out), nil
-	default:
-		// Non-castable type pair with at least one
-		// (non-record) container: output column is left
-		// unchanged.
-		return inType, nil
 	}
+	if bestUnionSelector(specType, inType) > -1 {
+		return specType, nil
+	}
+	// Non-castable type pair with at least one
+	// (non-record) container: output column is left
+	// unchanged.
+	return inType, nil
+}
+
+// bestUnionSelector tries to return the most specific selector for t in
+// maybeUnion.  It returns -1 if maybeUnion is not a union or contains no type
+// compatible with t.  (Types are compatible if they have the same underlying
+// type.)  If maybeUnion contains t, bestUnionSelector returns its selector.
+// Otherwise, if maybeUnion contains t's underlying type, bestUnionSelector
+// returns its selector.  Finally, bestUnionSelector returns the smallest
+// maybeUnion selector whose type is compatible with t.
+func bestUnionSelector(maybeUnion, t zng.Type) int {
+	u, ok := zng.AliasOf(maybeUnion).(*zng.TypeUnion)
+	if !ok {
+		return -1
+	}
+	aliasOfT := zng.AliasOf(t)
+	underlying := -1
+	compatible := -1
+	for i, ut := range u.Types {
+		if ut == t {
+			return i
+		}
+		if ut == aliasOfT && underlying == -1 {
+			underlying = i
+		}
+		if zng.AliasOf(ut) == aliasOfT && compatible == -1 {
+			compatible = i
+		}
```

Comment:
Not that we have to do this now, but you'd probably here want it to also have the ability to coerce a primitive if there's no exact matching type (for instance string and bstring), no? Feels like we don't do this consistently enough.

Reply:
That sounds intuitive, but the decision is complicated by the fact that it probably isn't the behavior we want for the `fuse` operator.